### PR TITLE
Adds a new colorbox element

### LINKED ISSUE #6510: Color selection is gone for Server Group
- Azure Data Studio Version:
```
Version: 1.10.0-insider (system setup)
Commit: 9f342ec9d5ed3827ecd86d7e165044e1b73e77ea
Date: 2019-07-26T07:49:17.517Z
VS Code 1.37.0
Electron: 4.2.5
Chrome: 69.0.3497.128
Node.js: 10.11.0
V8: 6.9.427.31-electron.0
OS: Windows_NT x64 10.0.18362
```

Steps to Reproduce:

1. Edit a current server group/Create a new server group
2. Unable to see color differences

This applies to dark and light theme, same result. Note these are **built-in themes** for ADS and not VS Code based extensions manually installed.

Light theme:
image

Dark theme:
image

### PR DESCRIPTION
**Pull request by aaomidi**
Fixes #6510 

This colorbox element is a new element. It's specifically designed for group colors.

Internally it uses a radio input type - which inherently enforces a single selection for the desired option. It also allows adding more custom functionality over the included VSCode checkbox/radio element.

### CODE REVIEW

File: src/sql/base/browser/ui/colorbox/colorbox.ts [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+		if (styles.backgroundColor) {
+			this.backgroundColor = styles.backgroundColor;
+		}
+		this.updateStyle();
+	}
+
+	private updateStyle(): void {
+		this.domNode.style.setProperty('--colorbox-bg', this.backgroundColor ? this.backgroundColor.toString() : '');
+		//this.domNode.style.setProperty('--colorbox-bg-border', this.backgroundColor ? this.backgroundColor.darken(0.2).toString() : '');
+
```
Comment (Charles-Gagnon) [inline src/sql/base/browser/ui/colorbox/colorbox.ts]: remove

**Comment by kburtram** [comment]
@aaomidi could you please add a screenshot of what the dialog looks like with this change?

File: src/sql/base/browser/ui/colorbox/colorbox.ts [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+		console.log(this.domNode.style.getPropertyValue('--colorbox-bg'));
+	}
+
+	public get checked(): boolean {
+		return this._checked;
+	}
+
+	public set checked(checked: boolean) {
+		this._checked = checked;
+		this.domNode.setAttribute('sql-colorbox-checked', String(this.checked));
```
Comment (Charles-Gagnon) [inline src/sql/base/browser/ui/colorbox/colorbox.ts]: >sql-colorbox-checked [](start = 29, length = 20)

What's this attribute used for?

File: src/sql/workbench/parts/objectExplorer/browser/serverGroupDialog.ts [hunk 3]
```diff
@@ -309,13 +310,13 @@ export class ServerGroupDialog extends Modal {
 	// update UI elements that have derivative behaviors based on other state changes
 	private updateView(): void {
 		// check the color buttons and if their checked state does not match the view model state then correct it
-		for (let i = 0; i < this._colorCheckBoxesMap.length; i++) {
-			let { checkbox, color } = this._colorCheckBoxesMap[i];
-			if ((this._viewModel.groupColor === color) && (checkbox.checked === false)) {
-				checkbox.checked = true;
+		for (let i = 0; i < this._colorColorBoxesMap.length; i++) {
+			let { colorbox: colorbox, color } = this._color	esMap[i];
```
Comment (Charles-Gagnon) [inline src/sql/workbench/parts/objectExplorer/browser/serverGroupDialog.ts]: >	 [](start = 50, length = 1)

It looks like a typo here - there's a tab where I assume there should be _colorColorBoxes

**Comment by aaomidi** [comment]
Video of the feature: https://i.imgur.com/n24Ne3G.mp4

Screenshot: 

image

**Comment by aaomidi** [comment]
TODO:
- [ ] aria-label for accessibility.

File: src/sql/base/browser/ui/colorbox/colorbox.ts (hunk 2 above)
Comment (aaomidi) [reply src/sql/base/browser/ui/colorbox/colorbox.ts]: Ah - that's from some WIP. I'll remove it now!

File: src/sql/base/browser/ui/colorbox/colorbox.ts [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+	public style(styles: ColorboxStyle): void {
+		if (styles.backgroundColor) {
+			this.backgroundColor = styles.backgroundColor;
+		}
+		this.updateStyle();
+	}
+
+	private updateStyle(): void {
+		this.domNode.style.setProperty('--colorbox-bg', this.backgroundColor ? this.backgroundColor.toString() : '');
```
Comment (anthonydresser) [inline src/sql/base/browser/ui/colorbox/colorbox.ts]: you can just set the background directly `this.domNode.syle.backgroundColor`

File: src/sql/base/browser/ui/colorbox/media/colorbox.css [hunk 5]
```diff
@@ -0,0 +1,15 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the Source EULA. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+.colorbox {
+	background: var(--colorbox-bg);
```
Comment (anthonydresser) [inline src/sql/base/browser/ui/colorbox/media/colorbox.css]: don't set colors in css, it should be programmic.

Comment (aaomidi) [reply src/sql/base/browser/ui/colorbox/media/colorbox.css]: Curious to know why? Is it just a codestyle for ADS or some other underlying reason?

Comment (anthonydresser) [reply src/sql/base/browser/ui/colorbox/media/colorbox.css]: colors should be theme aware, and using variables is just unnecessary and make the code harder to follow.

Comment (aaomidi) [reply src/sql/base/browser/ui/colorbox/media/colorbox.css]: Thanks that makes a lot of sense!

**Comment by Charles-Gagnon** [comment]
FYI - a lot of us use [Screen To Gif](https://www.screentogif.com/) for embedding gifs directly into PRs and other stuff - it's really nice for stuff like this

**Comment by aaomidi** [comment]
Would like to potentially look into using something like these for accessibility purposes:

- https://www.npmjs.com/package/cherangi
- https://www.npmjs.com/package/ntcjs

**Pull request merged**